### ISSUE TITLE
Enhancement: Add support for https healthMonitor protocol to backend property AND associate a client SSL profile to a health monitor.

### ISSUE THREAD

**Issue by Rodalli** [F5Networks/k8s-bigip-ctlr#942]
#### Description
Currently, the backend property for F5 Resources schema only supports three protocols: http, tcp and udp. There is no solution for monitoring https backends, or a use case in which the administrator may want to create a health monitor which leverages a client SSL profile (with a Certificate Authority configured) to perform an TLS handshake and validate certificates of backend web servers.

#### Kubernetes Version
1.12.3

#### Controller Version
1.9

#### BIG-IP Version
12.1.3

#### Diagnostic Information
As an example, this ConfigMap describing a virtual server configuration will fail if the backend server is only accessible via HTTPS:

```
kind: ConfigMap
apiVersion: v1
metadata:
  name: some-service-f5.vs
  labels:
    f5type: virtual-server
data:
  # https://clouddocs.f5.com/containers/latest/releases_and_versioning.html#f5-schema
  schema: "f5schemadb://bigip-virtual-server_v0.1.7.json"
  data: |
    {
      "virtualServer": {
        "backend": {
          "servicePort": 8443,
          "serviceName": "some-service",
          "healthMonitors": [{
            "interval": 5,
            "protocol": "http",
            "send": "HEAD / HTTP1.1",
            "recv": "200 OK",
            "timeout": 15
          }]
        },
        "frontend": {
          "virtualAddress": {
            "port": 443,
            "bindAddr": "192.168.10.1"
          },
          "partition": "k8s-ctlr",
          "balance": "least-connections-member"
        }
      }
    }
```

The only alternative seems to be to simply use a TCP health monitor, which is obviously not going to be very useful when you are using NodePort Service types and want the BIG-IP to stop sending traffic to a Pod that has failed. The exposed TCP port on the Kubernetes worker is going to remain open while the Pod is terminating and a new one is being spun up, and the BIG-IP will keep sending requests to it even though there's no actual HTTP server there to respond.

What are we to do if we need to monitor an HTTPS backend with its own certificate signed by a private CA (that we trust and have configured a SSL profile for on the BIG-IP)?

**Comment by mdditt2000** [comment]
Hi Michael, You can accomplish what you want todo by simply moving to AS3. AS3 is a declarative API shipping now for over a year with lots of production customer. We implemented AS3 in CIS 1.9.0 and made some enhancement is CIS 1.9.1. 

Here is the example you want.
https://github.com/mdditt2000/kubernetes/blob/dev/cis-1-9/as3%20declaration/end-to-end-ssl.yaml

I can be available tomorrow for a demo of the following and help you make this modification. Please let me know. Thanks Mark

**Comment by Rodalli** [comment]
@mdditt2000, looking at the example provided, if the backend servers in the pool are presenting an SSL certificate that's signed by an unknown CA, how can we make sure that the Monitor configured in the AS3 declaration trusts the CA that signed the backend server's certificate?

Looking at the AS3 specification for the Monitor class (https://clouddocs.f5.com/products/extensions/f5-appsvcs-extension/latest/refguide/schema-reference.html#monitor), I see that there is a clientCertificate property to use when authenticating to the pool members, but I don't see any property that allows us to specify a profile with a CA certificate to trust.

**Comment by mdditt2000** [comment]
@Rodalli Michael you are correct regarding AS3. AS3 can provide any BIG-IP needed. We also have a github page for AS3 questions or filing issues. https://github.com/F5Networks/f5-appsvcs-extension/issues. I am looking into the issue with @focrensh

**Comment by mdditt2000** [comment]
@Rodalli My colleagues was doing some testing. We created a internal AS3 issue for tracking **AUTOTOOL-450**

If you point the monitor at a cert directly, AS3 runs

  "testmon": {
    "class": "Monitor",
    "monitorType": "https",
    "clientCertificate": "webcert"
  },
  "webcert": {
    "class": "Certificate",
    "certificate": {
      "bigip": "/Common/default.crt"
    },
    "privateKey": {
      "bigip": "/Common/default.key"
    }
  }

In the UI/TMSH the monitor type accepts a profile as input, not a cert. If you try this with AS3 it throws an error:

"testmon": {
"class": "Monitor",
"monitorType": "https",
"clientCertificate": "webtls"
},
"webtls": {
"class": "TLS_Client",
"trustCA": {"bigip":"/Common/default.crt"}
},

"errors": [
"/test123/MyWebApp1/testmon/clientCertificate: **AS3 pointer webtls does not point to required object type"**
 
When the monitor references the cert, its requiring there to be a key also, which TMOS does not. Since BIG-IP is the client here, it should not need a key. This only is required if the monitor references it.

 "testmon": {
"class": "Monitor",
"monitorType": "https",
"clientCertificate": "webcert"
},
"webcert": {
"class": "Certificate",
"certificate": {"bigip":"/Common/default.crt"}
}

"message": "declaration failed",
"response": "0107134a:3: **File object by name (/test123/MyWebApp1/webcert.key) is missing**.",

**Comment by Rodalli** [comment]
@mdditt2000 Thanks for the testing effort and information...I was heading down a similar path as your second example, but now I can see from your testing that that's not going to work.

Thanks for filing the bug as well, as I agree that the client should not require a private key when you're not performing mutual authentication via TLS.

While waiting for a fix, would the workaround then be to provide a private key in the Certificate or TLS_Client object, even though it's not really needed?

**Comment by focrensh** [comment]
Even if the key is provided, it does not seem to populate the SSL profile within the monitor currently. This is also being tracked as part of the bug.

**Comment by Rodalli** [comment]
Thank you @focrensh and @mdditt2000. As it looks like there's currently no workaround, should we open a support ticket with F5 to try to get a fix for AUTOTOOL-450 prioritized? We have a pipeline in development right now that needs this functionality to deploy configurations to our BIG-IP devices.

**Comment by focrensh** [comment]
It his high priority currently. As it is, it will be included within the next release of AS3.

**Comment by mdditt2000** [comment]
@Rodalli here is the documentation. https://cloudsolutions.pages.gitswarm.f5net.com/f5-appsvcs/declarations/health-monitors.html#using-certificates-in-an-https-monitor

The issue has being resolved in the 3.13.0-1 Build. I can make the build available if you want to test.

**Comment by mdditt2000** [comment]
Issue has being resolved 3.13.0-1. AS3-13 will be published later this month. Please update the AS3 RPM on the BIP-IP. Let me know if you have any additional questions. Closing this issue out as resolved.
